src/backend/distributed/master/master_repair_shards.c

```diff
@@ -377,6 +386,63 @@ CopyShardCommandList(ShardInterval *shardInterval,
 
 
 /*
+ * CopyShardForeignConstraintCommandList generates command list to create foreign
+ * constraints existing in source shard after copying it to the other node.
+ */
+static List *
+CopyShardForeignConstraintCommandList(ShardInterval *shardInterval)
+{
+	List *copyShardForeignConstraintCommandList = NIL;
+
+	Oid schemaId = get_rel_namespace(shardInterval->relationId);
+	char *schemaName = get_namespace_name(schemaId);
+	char *escapedSchemaName = quote_literal_cstr(schemaName);
+	int shardIndex = 0;
+
+	List *commandList = GetTableForeignConstraintCommands(shardInterval->relationId);
+	ListCell *commandCell = NULL;
+
+	/* we will only use shardIndex if there is a foreign constraint */
+	if (commandList != NIL)
+	{
+		shardIndex = FindShardIntervalIndex(shardInterval);
+	}
+
+	foreach(commandCell, commandList)
+	{
+		char *command = (char *) lfirst(commandCell);
+		char *escapedCommand = quote_literal_cstr(command);
+
+		Oid referencedRelationId = InvalidOid;
+		Oid referencedSchemaId = InvalidOid;
+		char *referencedSchemaName = NULL;
+		char *escapedReferencedSchemaName = NULL;
+		uint64 referencedShardId = INVALID_SHARD_ID;
+
+		StringInfo applyForeignConstraintCommand = makeStringInfo();
+
+		/* we need to parse the foreign constraint command to get referencing table id */
+		referencedRelationId = ForeignConstraintGetReferencedTableId(command);
+		referencedSchemaId = get_rel_namespace(referencedRelationId);
```

Comment:
Should we error out here if referencedRelationId is InvalidOid (as a friendly form of assert)?